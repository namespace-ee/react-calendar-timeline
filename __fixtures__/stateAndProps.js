import { defaultKeys } from 'lib/default-config'
import {items} from './itemsAndGroups'

export const visibleTimeStart = 1540501200000
export const visibleTimeEnd = 1540587600000

export const props = {
  keys: defaultKeys,
  lineHeight: 30,
  stackItems: true,
  itemHeightRatio: 0.75,
  visibleTimeEnd,
  visibleTimeStart,
}

export const propsNoStack = {
  ...props,
  stackItems: false,
}


export const state = {
  draggingItem: undefined,
  dragTime: null,
  resizingItem: null,
  resizingEdge: null,
  resizeTime: null,
  newGroupOrder: null,
  canvasTimeStart: 1540414800000,
  visibleTimeEnd: visibleTimeEnd,
  visibleTimeStart: visibleTimeStart,
  canvasTimeEnd: 1540674000000,
  width: 1000,
}

//offset 1 hour
const timeOffset = 1 * 60 *60 *1000

export const stateMoveItem = {
  ...state,
  draggingItem: items[0].id,
  dragTime: items[0].start_time+timeOffset,
  newGroupOrder: 0,
}
export const stateResizeItemLeft = {
  ...state,
  resizingItem: items[0].id,
  resizingEdge: 'left',
  resizeTime: items[0].start_time+timeOffset,
  newGroupOrder: 0,
}

export const stateResizeItemRight = {
  ...state,
  resizingItem: items[0].id,
  resizingEdge: 'right',
  resizeTime: items[0].end_time+timeOffset,
  newGroupOrder: 0,
}