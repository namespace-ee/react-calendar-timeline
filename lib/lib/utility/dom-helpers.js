"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getParentPosition = getParentPosition;
exports.getSumScroll = getSumScroll;
exports.getSumOffset = getSumOffset;

// TODO: can we use getBoundingClientRect instead??
// last place this is used is in "handleWheel" in ScrollElement
function getParentPosition(element) {
  var xPosition = 0;
  var yPosition = 0;
  var first = true;

  while (element) {
    if (!element.offsetParent && element.tagName === 'BODY' && element.scrollLeft === 0 && element.scrollTop === 0) {
      element = document.scrollingElement || element;
    }

    xPosition += element.offsetLeft - (first ? 0 : element.scrollLeft) + element.clientLeft;
    yPosition += element.offsetTop - (first ? 0 : element.scrollTop) + element.clientTop;
    element = element.offsetParent;
    first = false;
  }

  return {
    x: xPosition,
    y: yPosition
  };
}

function getSumScroll(node) {
  if (node === document.body) {
    return {
      scrollLeft: 0,
      scrollTop: 0
    };
  } else {
    var parent = getSumScroll(node.parentNode);
    return {
      scrollLeft: node.scrollLeft + parent.scrollLeft,
      scrollTop: node.scrollTop + parent.scrollTop
    };
  }
}

function getSumOffset(node) {
  if (node === document.body || !node.offsetParent) {
    return {
      offsetLeft: 0,
      offsetTop: 0
    };
  } else {
    var parent = getSumOffset(node.offsetParent);
    return {
      offsetLeft: node.offsetLeft + parent.offsetLeft,
      offsetTop: node.offsetTop + parent.offsetTop
    };
  }
}