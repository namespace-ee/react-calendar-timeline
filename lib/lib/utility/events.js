"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.composeEvents = composeEvents;

function composeEvents() {
  for (var _len = arguments.length, fns = new Array(_len), _key = 0; _key < _len; _key++) {
    fns[_key] = arguments[_key];
  }

  return function (event) {
    for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
      args[_key2 - 1] = arguments[_key2];
    }

    event.preventDefault();
    fns.forEach(function (fn) {
      return fn && fn.apply(void 0, [event].concat(args));
    });
  };
}