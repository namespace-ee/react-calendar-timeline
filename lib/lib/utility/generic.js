"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._get = _get;
exports._length = _length;
exports.arraysEqual = arraysEqual;
exports.deepObjectCompare = deepObjectCompare;
exports.keyBy = keyBy;
exports.noop = noop;

var _lodash = _interopRequireDefault(require("lodash.isequal"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

// so we could use both immutable.js objects and regular objects
function _get(object, key) {
  return typeof object.get === 'function' ? object.get(key) : object[key];
}

function _length(object) {
  return typeof object.count === 'function' ? object.count() : object.length;
}

function arraysEqual(array1, array2) {
  return _length(array1) === _length(array2) && array1.every(function (element, index) {
    return element === _get(array2, index);
  });
}

function deepObjectCompare(obj1, obj2) {
  return (0, _lodash["default"])(obj1, obj2);
}

function keyBy(value, key) {
  var obj = {};
  value.forEach(function (element) {
    obj[element[key]] = element;
  });
  return obj;
}

function noop() {}