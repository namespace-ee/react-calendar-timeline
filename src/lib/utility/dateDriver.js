import { vanillaDateDriver } from "./dateDrivers/vanillaDateDriver";
// If someone overwrites this, they wanted to.
const obscureKey = "reactCalendarTimelineDateDriver";

if (!window[obscureKey]) {
  window[obscureKey] = (function () {
    let driver = vanillaDateDriver;
    const service = {
      get: () => {
        return driver
      },
      set: (driver) => {
        return driver;
      }
    };
    Object.freeze(service);
    return service;
  })();
}

export const setDateDriver = window[obscureKey].set;
export const dateDriver = window[obscureKey].get();
