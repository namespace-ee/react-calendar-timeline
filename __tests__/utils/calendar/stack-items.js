import { stackTimelineItems } from 'lib/utility/calendar'
import { items, groups } from '../../../__fixtures__/itemsAndGroups'
import {
  props,
  state,
  stateMoveItem,
  stateResizeItemLeft,
  stateResizeItemRight,
  propsNoStack,
} from '../../../__fixtures__/stateAndProps'
describe('stackItems', () => {
  it('work as expected', () => {
    expect(
      stackTimelineItems(
        items,
        groups,
        9000,
        state.canvasTimeStart,
        state.canvasTimeEnd,
        props.keys,
        props.minHeight,
        props.lineHeight,
        props.itemHeightRatio,
        props.stackItems,
        state.draggingItem,
        state.resizingItem,
        state.dragTime,
        state.resizingEdge,
        state.resizeTime,
        state.newGroupOrder
      )
    ).toMatchSnapshot()
  })
  it('work as expected no stack', () => {
    expect(
      stackTimelineItems(
        items,
        groups,
        9000,
        state.canvasTimeStart,
        state.canvasTimeEnd,
        propsNoStack.keys,
        propsNoStack.minHeight,
        propsNoStack.lineHeight,
        propsNoStack.itemHeightRatio,
        propsNoStack.stackItems,
        state.draggingItem,
        state.resizingItem,
        state.dragTime,
        state.resizingEdge,
        state.resizeTime,
        state.newGroupOrder
      )
    ).toMatchSnapshot()
  })
  it('should stack items while moving an item', () => {
    expect(
      stackTimelineItems(
        items,
        groups,
        9000,
        stateMoveItem.canvasTimeStart,
        stateMoveItem.canvasTimeEnd,
        props.keys,
        props.minHeight,
        props.lineHeight,
        props.minHeight,
        props.itemHeightRatio,
        props.stackItems,
        stateMoveItem.draggingItem,
        stateMoveItem.resizingItem,
        stateMoveItem.dragTime,
        stateMoveItem.resizingEdge,
        stateMoveItem.resizeTime,
        stateMoveItem.newGroupOrder
      )
    ).toMatchSnapshot()
  })
  it('should stack items while resize item left', () => {
    expect(
      stackTimelineItems(
        items,
        groups,
        9000,
        stateResizeItemLeft.canvasTimeStart,
        stateResizeItemLeft.canvasTimeEnd,
        props.keys,
        props.minHeight,
        props.lineHeight,
        props.itemHeightRatio,
        props.stackItems,
        stateResizeItemLeft.draggingItem,
        stateResizeItemLeft.resizingItem,
        stateResizeItemLeft.dragTime,
        stateResizeItemLeft.resizingEdge,
        stateResizeItemLeft.resizeTime,
        stateResizeItemLeft.newGroupOrder
      )
    ).toMatchSnapshot()
  })
  it('should stack items while resize item right', () => {
    expect(
      stackTimelineItems(
        items,
        groups,
        9000,
        stateResizeItemRight.canvasTimeStart,
        stateResizeItemRight.canvasTimeEnd,
        props.keys,
        props.minHeight,
        props.lineHeight,
        props.itemHeightRatio,
        props.stackItems,
        stateResizeItemRight.draggingItem,
        stateResizeItemRight.resizingItem,
        stateResizeItemRight.dragTime,
        stateResizeItemRight.resizingEdge,
        stateResizeItemRight.resizeTime,
        stateResizeItemRight.newGroupOrder
      )
    ).toMatchSnapshot()
  })
  it('should return empty dimensions if groups are empty', () => {
    expect(
      stackTimelineItems(
        items,
        [],
        9000,
        state.canvasTimeStart,
        state.canvasTimeEnd,
        props.keys,
        props.minHeight,
        props.lineHeight,
        props.itemHeightRatio,
        props.stackItems,
        state.draggingItem,
        state.resizingItem,
        state.dragTime,
        state.resizingEdge,
        state.resizeTime,
        state.newGroupOrder
      )
    ).toMatchObject({
      dimensionItems: [],
      height: 0,
      groupHeights: [],
      groupTops: []
    })
  })
})
