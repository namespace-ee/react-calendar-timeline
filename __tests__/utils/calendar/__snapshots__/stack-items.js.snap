// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stackItems should stack items while moving an item 1`] = `
Object {
  "dimensionItems": Array [
    Object {
      "dimensions": Object {
        "collisionLeft": 1540543600000,
        "collisionWidth": 6803877,
        "height": 22.5,
        "left": 4472.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 236.24572916666602,
      },
      "id": "0",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "height": 22.5,
        "left": 4097.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 33.75,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "height": 22.5,
        "left": 4722.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "height": 22.5,
        "left": 5388.888888888889,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 33.75,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 7125,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 8375,
        "order": Object {
          "group": Object {
            "id": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 93.75,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": Array [
    60,
    30,
    30,
  ],
  "groupTops": Array [
    0,
    60,
    90,
  ],
  "height": 120,
}
`;

exports[`stackItems should stack items while resize item left 1`] = `
Object {
  "dimensionItems": Array [
    Object {
      "dimensions": Object {
        "collisionLeft": 1540543600000,
        "collisionWidth": 3203877,
        "height": 22.5,
        "left": 4472.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 111.24572916666602,
      },
      "id": "0",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "height": 22.5,
        "left": 4097.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 33.75,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "height": 22.5,
        "left": 4722.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "height": 22.5,
        "left": 5388.888888888889,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 33.75,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 7125,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 8375,
        "order": Object {
          "group": Object {
            "id": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 93.75,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": Array [
    60,
    30,
    30,
  ],
  "groupTops": Array [
    0,
    60,
    90,
  ],
  "height": 120,
}
`;

exports[`stackItems should stack items while resize item right 1`] = `
Object {
  "dimensionItems": Array [
    Object {
      "dimensions": Object {
        "collisionLeft": 1540540000000,
        "collisionWidth": 10403877,
        "height": 22.5,
        "left": 4347.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 361.245729166666,
      },
      "id": "0",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "height": 22.5,
        "left": 4097.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 33.75,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "height": 22.5,
        "left": 4722.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "height": 22.5,
        "left": 5388.888888888889,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 33.75,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 7125,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 8375,
        "order": Object {
          "group": Object {
            "id": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 93.75,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": Array [
    60,
    30,
    30,
  ],
  "groupTops": Array [
    0,
    60,
    90,
  ],
  "height": 120,
}
`;

exports[`stackItems work as expected 1`] = `
Object {
  "dimensionItems": Array [
    Object {
      "dimensions": Object {
        "collisionLeft": 1540540000000,
        "collisionWidth": 6803877,
        "height": 22.5,
        "left": 4347.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 236.24572916666602,
      },
      "id": "0",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "height": 22.5,
        "left": 4097.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 33.75,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "height": 22.5,
        "left": 4722.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "height": 22.5,
        "left": 5388.888888888889,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 33.75,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 7125,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 8375,
        "order": Object {
          "group": Object {
            "id": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 93.75,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": Array [
    60,
    30,
    30,
  ],
  "groupTops": Array [
    0,
    60,
    90,
  ],
  "height": 120,
}
`;

exports[`stackItems work as expected no stack 1`] = `
Object {
  "dimensionItems": Array [
    Object {
      "dimensions": Object {
        "collisionLeft": 1540540000000,
        "collisionWidth": 6803877,
        "height": 22.5,
        "left": 4347.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 236.24572916666602,
      },
      "id": "0",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540532800000,
        "collisionWidth": 21203877,
        "height": 22.5,
        "left": 4097.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 736.245729166666,
      },
      "id": "5",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540550800000,
        "collisionWidth": 24803877,
        "height": 22.5,
        "left": 4722.222222222223,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 861.245729166666,
      },
      "id": "6",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540570000000,
        "collisionWidth": 14875919,
        "height": 22.5,
        "left": 5388.888888888889,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 516.5249652777784,
      },
      "id": "1",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540620000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 7125,
        "order": Object {
          "group": Object {
            "id": "1",
          },
          "index": 0,
        },
        "stack": true,
        "top": 3.75,
        "width": 708.2481944444444,
      },
      "id": "2",
    },
    Object {
      "dimensions": Object {
        "collisionLeft": 1540656000000,
        "collisionWidth": 20397548,
        "height": 22.5,
        "left": 8375,
        "order": Object {
          "group": Object {
            "id": "3",
          },
          "index": 2,
        },
        "stack": true,
        "top": 63.75,
        "width": 625,
      },
      "id": "3",
    },
  ],
  "groupHeights": Array [
    30,
    30,
    30,
  ],
  "groupTops": Array [
    0,
    30,
    60,
  ],
  "height": 90,
}
`;
